In [2919]:
import pandas as pd
import numpy as np
import os

In [2920]:
#Get verifica.ipynb notebook functions
%run verifica.ipynb

# Pre-Processing Radiologists' Data

#### Load Data from excel file

In [2921]:

file_path = '/Users/larasousa/Desktop/uni/radiologistas.xls'
df_radiologists = pd.read_excel(file_path)
df_radiologists

TCIA Patient ID  \
0    LIDC-IDRI-0068   
1    LIDC-IDRI-0071   
2    LIDC-IDRI-0072   
3    LIDC-IDRI-0088   
4    LIDC-IDRI-0090   
..              ...   
152  LIDC-IDRI-0994   
153  LIDC-IDRI-1002   
154  LIDC-IDRI-1004   
155  LIDC-IDRI-1010   
156  LIDC-IDRI-1011   

     Diagnosis at the Patient Level\n0=Unknown\n1=benign or non-malignant disease\n2= malignant, primary lung cancer\n3 = malignant metastatic\n  \
0                                                    3                                                                                             
1                                                    3                                                                                             
2                                                    2                                                                                             
3                                                    3                                                                                             
4                                                    2                                                                                             
..                                                 ...                                                                                             
152                                                  2                                                                                             
153                                                  2                                                                                             
154                                                  2                                                                                             
155                                                  0                                                                                             
156                                                  3                                                                                             

     Diagnosis Method\n0 = unknown\n1 = review of radiological images to show 2 years of stable nodule\n2 = biopsy\n3 = surgical resection\n4 = progression or response  \
0                                                    4                                                                                                                    
1                                                    1                                                                                                                    
2                                                    4                                                                                                                    
3                                                    0                                                                                                                    
4                                                    3                                                                                                                    
..                                                 ...                                                                                                                    
152                                                  3                                                                                                                    
153                                                  2                                                                                                                    
154                                                  3                                                                                                                    
155                                                  0                                                                                                                    
156                                                  2               

### Rename Columns

In [2922]:

df_radiologists = df_radiologists.rename(columns={
    'Diagnosis at the Patient Level\n0=Unknown\n1=benign or non-malignant disease\n2= malignant, primary lung cancer\n3 = malignant metastatic\n': 'Diagnosis at the patient',
    'Diagnosis Method\n0 = unknown\n1 = review of radiological images to show 2 years of stable nodule\n2 = biopsy\n3 = surgical resection\n4 = progression or response': 'Diagnosis Method',
    'Nodule 1\nDiagnosis at the Nodule Level \n0=Unknown\n1=benign or non-malignant disease\n2= malignant, primary lung cancer\n3 = malignant metastatic)\n':'Nodule 1 Diagnosis at the Nodule Level',
    'Nodule 1\nDiagnosis Method at the Nodule Level\n0 = unknown\n1 = review of radiological images to show 2 years of stable nodule\n2 = biopsy\n3 = surgical resection\n4 = progression or response\n':'Nodule 1 Diagnosis Method at the Nodule Level',
    'Nodule 2\nDiagnosis at the Nodule Level \n0=Unknown\n1=benign or non-malignant disease\n2= malignant, primary lung cancer\n3 = malignant metastatic)\n':'Nodule 2 Diagnosis at the Nodule Level',
    'Nodule 2\nDiagnosis Method at the Nodule Level\n0 = unknown\n1 = review of radiological images to show 2 years of stable nodule\n2 = biopsy\n3 = surgical resection\n4 = progression or response\n':'Nodule 2 Diagnosis Method at the Nodule Level',
    'Nodule 3\nDiagnosis at the Nodule Level \n0=Unknown\n1=benign or non-malignant disease\n2= malignant, primary lung cancer\n3 = malignant metastatic)\n':'Nodule 3 Diagnosis at the Nodule Level',
    'Nodule 3\nDiagnosis Method at the Nodule Level\n0 = unknown\n1 = review of radiological images to show 2 years of stable nodule\n2 = biopsy\n3 = surgical resection\n4 = progression or response\n':'Nodule 3 Diagnosis Method at the Nodule Level',
    'Nodule 4\nDiagnosis at the Nodule Level \n0=Unknown\n1=benign or non-malignant disease\n2= malignant, primary lung cancer\n3 = malignant metastatic)\n':'Nodule 4 Diagnosis at the Nodule Level',
    'Nodule 4\nDiagnosis Method at the Nodule Level\n0 = unknown\n1 = review of radiological images to show 2 years of stable nodule\n2 = biopsy\n3 = surgical resection\n4 = progression or response\n':'Nodule 4 Diagnosis Method at the Nodule Level',
    'Nodule 5\nDiagnosis at the Nodule Level \n0=Unknown\n1=benign or non-malignant disease\n2= malignant, primary lung cancer\n3 = malignant metastatic)\n': 'Nodule 5 Diagnosis at the Nodule Level',
    'Nodule 5\nDiagnosis Method at the Nodule Level\n0 = unknown\n1 = review of radiological images to show 2 years of stable nodule\n2 = biopsy\n3 = surgical resection\n4 = progression or response\n':'Nodule 5 Diagnosis Method at the Nodule Level',
})
df_radiologists

TCIA Patient ID  Diagnosis at the patient  Diagnosis Method  \
0    LIDC-IDRI-0068                         3                 4   
1    LIDC-IDRI-0071                         3                 1   
2    LIDC-IDRI-0072                         2                 4   
3    LIDC-IDRI-0088                         3                 0   
4    LIDC-IDRI-0090                         2                 3   
..              ...                       ...               ...   
152  LIDC-IDRI-0994                         2                 3   
153  LIDC-IDRI-1002                         2                 2   
154  LIDC-IDRI-1004                         2                 3   
155  LIDC-IDRI-1010                         0                 0   
156  LIDC-IDRI-1011                         3                 2   

    Primary tumor site for metastatic disease  \
0                          Head & Neck Cancer   
1                                Head & Neck    
2                                 Lung Cancer   
3                              Uterine Cancer   
4                                       NSCLC   
..                                        ...   
152                         LUL Large cell CA   
153                  non-small cell carcinoma   
154                                 LUL NSCLC   
155                                  lymphoma   
156        small-cell carcinoma of the tongue   

     Nodule 1 Diagnosis at the Nodule Level  \
0                                       3.0   
1                                       1.0   
2                                       1.0   
3                                       0.0   
4                                       2.0   
..                                      ...   
152                                     2.0   
153                                     NaN   
154                                     2.0   
155                                     0.0   
156                                     3.0   

     Nodule 1 Diagnosis Method at the Nodule Level  \
0                                              4.0   
1                                              1.0   
2                                              4.0   
3                                              0.0   
4                                              3.0   
..                                             ...   
152                                            3.0   
153                                            NaN   
154                                            3.0   
155                                            0.0   
156                                            2.0   

     Nodule 2 Diagnosis at the Nodule Level  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   
..                                      ...   
152                                     NaN   
153                                     NaN   
154                                     NaN   
155                                     NaN   
156                                     3.0   

     Nodule 2 Diagnosis Method at the Nodule Level  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   
..                                             ...   
152                                            NaN   
153                                            NaN   
154                                            NaN   
155                                            NaN   
156                                            4.0   

    Nodule 3 Diagnosis at the Nodule Level  \
0                                      NaN   
1                                      NaN   
2                                      NaN  

#### Radiologists' Data Verification

In [2923]:
verif_format_ids(df_radiologists, 'TCIA Patient ID')

The format of IDs in column 'TCIA Patient ID' is the same (LIDC-IDRI- followed by a four-digit number).


In [2924]:
duplicated(df_radiologists, 'TCIA Patient ID')

No duplicated IDs found in column 'TCIA Patient ID'.


In [2925]:
missing_data(df_radiologists)

Missing values in TCIA Patient ID: 0
Missing values in Diagnosis at the patient: 0
Missing values in Diagnosis Method: 0
Missing values in Primary tumor site for metastatic disease: 0
Missing values in Nodule 1 Diagnosis at the Nodule Level: 20
Missing values in Nodule 1 Diagnosis Method at the Nodule Level: 20
Missing values in Nodule 2 Diagnosis at the Nodule Level: 137
Missing values in Nodule 2 Diagnosis Method at the Nodule Level: 137
Missing values in Nodule 3 Diagnosis at the Nodule Level: 156
Missing values in Nodule 3 Diagnosis Method at the Nodule Level: 156
Missing values in Nodule 4 Diagnosis at the Nodule Level: 157
Missing values in Nodule 4 Diagnosis Method at the Nodule Level: 157
Missing values in Nodule 5 Diagnosis at the Nodule Level: 157
Missing values in Nodule 5 Diagnosis Method at the Nodule Level: 157


In [2926]:
unique_values = df_radiologists['Diagnosis at the patient'].unique()
print(unique_values)


[3 2 1 0]


#### Processing Radiologists' Data

In [2927]:
df_radiologists = remove_unknown_rows(df_radiologists, 'Diagnosis at the patient', 0)


In [2928]:
#shape after removing unknown diagnosis
df_radiologists.shape

(130, 14)

In [2929]:
unique_primary_tumor_names = df_radiologists['Primary tumor site for metastatic disease'].unique()
unique_primary_tumor_names

array(['Head & Neck Cancer', 'Head & Neck ', 'Lung Cancer',
       'Uterine Cancer', 'NSCLC', 'urothelial carcinoma', 'Testis ',
       'Prostate ', 'colon cancer', 'colon', 'Metastatic colon cancer',
       ' Basaloid squamous carcinoma', 'osteosarcoma', 'chondrosarcoma',
       'thyroid carcinoma', 'pancreatic cancer', 'gallbladder',
       'non small cell lung', 'colorectal cancer', 'vaginal cancer',
       'stage III lung cancer', 'LGL (T cell lymphoma)',
       'organizing pneumonia, also leukemia',
       'granular cell tumor of the trachea', 'Lymphoma',
       'nonsmall cell lung cancer', 'melanoma',
       'adenoid cystic carcinoma', 'lung cancer',
       'Pulmonary cryptococcosis', 'Ewing sarcoma', 'HCC',
       'ovarian cancer', 'renal cell carcinoma', 'giant cell tumor',
       'metastatic rectal cancer', 'Colon', 'Small cell osteosarcoma',
       'testicular cancer ', 'non small cell lung, lymphoma',
       'Left chest wall chondrosarcoma', 'metastatic melanoma',
       're

In [2930]:
values_list = df_radiologists['Primary tumor site for metastatic disease'].tolist()
values_list

['Head & Neck Cancer',
 'Head & Neck ',
 'Lung Cancer',
 'Uterine Cancer',
 'NSCLC',
 'urothelial carcinoma',
 'Testis ',
 'Prostate ',
 'colon cancer',
 'colon',
 'Metastatic colon cancer',
 ' Basaloid squamous carcinoma',
 'Lung Cancer',
 'osteosarcoma',
 'chondrosarcoma',
 'thyroid carcinoma',
 'pancreatic cancer',
 'gallbladder',
 'non small cell lung',
 'colorectal cancer',
 'vaginal cancer',
 'stage III lung cancer',
 'LGL (T cell lymphoma)',
 'organizing pneumonia, also leukemia',
 'granular cell tumor of the trachea',
 'Lymphoma',
 'nonsmall cell lung cancer',
 'thyroid carcinoma',
 'melanoma',
 'nonsmall cell lung cancer',
 'LGL (T cell lymphoma)',
 'adenoid cystic carcinoma',
 'lung cancer',
 'Pulmonary cryptococcosis',
 'Ewing sarcoma',
 'HCC',
 'nonsmall cell lung cancer',
 'organizing pneumonia, also leukemia',
 'ovarian cancer',
 'renal cell carcinoma',
 'giant cell tumor',
 'lung cancer',
 'metastatic rectal cancer',
 'colon cancer',
 'lung cancer',
 'Colon',
 'Small cel

##### Simplify the names in Primary tumor site for metastatic disease column


In [2931]:

primary=['Head,  Neck',
'Head, Neck ',
'Lung',
'Uterine',
'Lung',
'Carcinoma',
'Prostate ',
'Prostate',
'Colon',
'Colon',
'Metastatic',
'Carcinoma',
'Lung',
'Sarcoma',
'Sarcoma',
'Carcinoma',
'Pancreatic',
'Bladder',
'Lung',
'Colon',
'Vaginal',
'Lung',
'Lymphoma',
'Leukemia',
'Trachea',
'Lymphoma',
'Lung',
'Carcinoma',
'Melanoma',
'Lung',
'Lymphoma',
'Carcinoma',
'Lung',
'Lung',
'Sarcoma',
'Carcinoma',
'Lung',
'Leukemia',
'Ovaries',
'Carcinoma',
'Bones',
'Lung',
'Metastatic',
'Colon',
'Lung',
'Colon',
'Sarcoma',
'Prostate',
'Lung, Lymphoma',
'Sarcoma',
'Metastatic',
'Rectum',
'Lung',
'Breast',
'Lung',
'Colon',
'Lung',
'Leukemia',
'Thyroid',
'Carcinoma',
'Sarcoma',
'Sarcoma',
'Carcinoid',
'Lung',
'Sarcoma',
'Lymphoma',
'Rectum',
'Lung',
'Laryngeal ',
'Carcinoma',
'Head,  Neck',
'Melanoma',
'Colon',
'Colon, Lung',
"Lymphoma ",
'Lung',
'Lung',
'Melanoma ',
'Lung',
'Breast',
'Lung',
'Sarcoma',
'Lymphoma',
'Lung',
'Melanoma',
'Breast',
'Breast, Carcinoma',
'Tonsils',
'Esophagus',
'multiple primary cancers',
'Lymphoma',
'Lymphoma',
'Melanoma',
'Sarcoma',
'solitary fibrous tumor, kidneys',
'Lung',
'Carcinoma',
'Carcinoma',
'Lung',
'Lung',
'Lung',
'Cervical',
'reactive mesothelial cells',
'Carcinoma',
'Lung',
'Lung',
'focal fibrosis',
'Carcinoma',
'Carcinoma',
'Brain',
'Carcinoma',
'Lung',
'Lung',
'Lung',
'Lung',
'Carcinoma',
'Carcinoma',
'carcinoma',
'Carcinoma',
'Carcinoma',
'Metastatic',
'Lung',
'Carcinoma',
'Carcinoma',
'Lung',
'Carcinoma',
'Lung',
'Carcinoma',
'Lung',
'Carcinoma']   
         
df_radiologists.shape

#substitui na coluna "Primary tumor site for metastatic disease" pelos nomes da lista "primary"
for i in range(len(df_radiologists)):
    if i < len(primary):
        df_radiologists.iloc[i, df_radiologists.columns.get_loc('Primary tumor site for metastatic disease')] = primary[i]

df_radiologists

TCIA Patient ID  Diagnosis at the patient  Diagnosis Method  \
0    LIDC-IDRI-0068                         3                 4   
1    LIDC-IDRI-0071                         3                 1   
2    LIDC-IDRI-0072                         2                 4   
3    LIDC-IDRI-0088                         3                 0   
4    LIDC-IDRI-0090                         2                 3   
..              ...                       ...               ...   
151  LIDC-IDRI-0965                         2                 3   
152  LIDC-IDRI-0994                         2                 3   
153  LIDC-IDRI-1002                         2                 2   
154  LIDC-IDRI-1004                         2                 3   
156  LIDC-IDRI-1011                         3                 2   

    Primary tumor site for metastatic disease  \
0                                 Head,  Neck   
1                                 Head, Neck    
2                                        Lung   
3                                     Uterine   
4                                        Lung   
..                                        ...   
151                                 Carcinoma   
152                                      Lung   
153                                 Carcinoma   
154                                      Lung   
156                                 Carcinoma   

     Nodule 1 Diagnosis at the Nodule Level  \
0                                       3.0   
1                                       1.0   
2                                       1.0   
3                                       0.0   
4                                       2.0   
..                                      ...   
151                                     NaN   
152                                     2.0   
153                                     NaN   
154                                     2.0   
156                                     3.0   

     Nodule 1 Diagnosis Method at the Nodule Level  \
0                                              4.0   
1                                              1.0   
2                                              4.0   
3                                              0.0   
4                                              3.0   
..                                             ...   
151                                            NaN   
152                                            3.0   
153                                            NaN   
154                                            3.0   
156                                            2.0   

     Nodule 2 Diagnosis at the Nodule Level  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   
..                                      ...   
151                                     NaN   
152                                     NaN   
153                                     NaN   
154                                     NaN   
156                                     3.0   

     Nodule 2 Diagnosis Method at the Nodule Level  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   
..                                             ...   
151                                            NaN   
152                                            NaN   
153                                            NaN   
154                                            NaN   
156                                            4.0   

    Nodule 3 Diagnosis at the Nodule Level  \
0                                      NaN   
1                                      NaN   
2                                      NaN  

In [2932]:
df_radiologists.columns

Index(['TCIA Patient ID', 'Diagnosis at the patient', 'Diagnosis Method',
       'Primary tumor site for metastatic disease',
       'Nodule 1 Diagnosis at the Nodule Level',
       'Nodule 1 Diagnosis Method at the Nodule Level',
       'Nodule 2 Diagnosis at the Nodule Level',
       'Nodule 2 Diagnosis Method at the Nodule Level',
       'Nodule 3 Diagnosis at the Nodule Level',
       'Nodule 3 Diagnosis Method at the Nodule Level',
       'Nodule 4 Diagnosis at the Nodule Level',
       'Nodule 4 Diagnosis Method at the Nodule Level',
       'Nodule 5 Diagnosis at the Nodule Level',
       'Nodule 5 Diagnosis Method at the Nodule Level'],
      dtype='object')

### Most Relevant Data

##### Store most relevant data in a Dataframe

In [2933]:

radiologists = pd.DataFrame(columns=["TCIA Patient ID", "Diagnosis at the patient", "Diagnosis Method", "Primary tumor site for metastatic disease"])
radiologists["TCIA Patient ID"] = df_radiologists["TCIA Patient ID"].copy()
radiologists["Diagnosis at the patient"] = df_radiologists["Diagnosis at the patient"].copy()
radiologists["Diagnosis Method"] = df_radiologists["Diagnosis Method"].copy()
radiologists["Primary tumor site for metastatic disease"] = df_radiologists["Primary tumor site for metastatic disease"].copy()


In [2934]:
radiologists.head()

TCIA Patient ID  Diagnosis at the patient  Diagnosis Method  \
0  LIDC-IDRI-0068                         3                 4   
1  LIDC-IDRI-0071                         3                 1   
2  LIDC-IDRI-0072                         2                 4   
3  LIDC-IDRI-0088                         3                 0   
4  LIDC-IDRI-0090                         2                 3   

  Primary tumor site for metastatic disease  
0                               Head,  Neck  
1                               Head, Neck   
2                                      Lung  
3                                   Uterine  
4                                      Lung

In [2935]:
radiologists.tail()

TCIA Patient ID  Diagnosis at the patient  Diagnosis Method  \
151  LIDC-IDRI-0965                         2                 3   
152  LIDC-IDRI-0994                         2                 3   
153  LIDC-IDRI-1002                         2                 2   
154  LIDC-IDRI-1004                         2                 3   
156  LIDC-IDRI-1011                         3                 2   

    Primary tumor site for metastatic disease  
151                                 Carcinoma  
152                                      Lung  
153                                 Carcinoma  
154                                      Lung  
156                                 Carcinoma

###### Discussion: Applying verif_format_ids funtion, it was verificated that the format is the same for all ids in TCIA Patient ID column, then with the duplicated function, it was concluded that there aren't duplicate ids. About the missing data, the function indicated that there are missing values, however its irrelevant because we stored just the important columns in a dataframe, and those columns don’t have ant missing values.

# Pre-Processing Nodule

#### Loading Nodule Dataset

In [2936]:

path = '/Users/larasousa/Desktop/uni/nodule.xlsx'
df_nodule = pd.read_excel(path)
df_nodule

TCIA Patent ID  Total Number of Nodules*   Number of Nodules >=3mm**  \
0     LIDC-IDRI-0001                          4                          1   
1     LIDC-IDRI-0002                         12                          1   
2     LIDC-IDRI-0003                          4                          4   
3     LIDC-IDRI-0004                          4                          1   
4     LIDC-IDRI-0005                          9                          3   
...              ...                        ...                        ...   
1014  LIDC-IDRI-1009                          2                          1   
1015  LIDC-IDRI-1010                         10                          1   
1016  LIDC-IDRI-1011                          4                          4   
1017  LIDC-IDRI-1012                          1                          1   
1018             NaN                       7371                       2669   

      Number of Nodules <3mm***  Unnamed: 4  \
0                             3         NaN   
1                            11         NaN   
2                             0         NaN   
3                             3         NaN   
4                             6         NaN   
...                         ...         ...   
1014                          1         NaN   
1015                          9         NaN   
1016                          0         NaN   
1017                          0         NaN   
1018                       4702         NaN   

                                             Unnamed: 5  
0                                                   NaN  
1     *total number of lesions that received either ...  
2     **total number of lesions that received a "nod...  
3     ***total number of lesions that received a "no...  
4                                                   NaN  
...                                                 ...  
1014                                                NaN  
1015                                                NaN  
1016                                                NaN  
1017                                                NaN  
1018                                                NaN  

[1019 rows x 6 columns]

In [2937]:
df_nodule = df_nodule.rename(columns={'TCIA Patent ID' : 'TCIA Patient ID'})
df_nodule

TCIA Patient ID  Total Number of Nodules*   Number of Nodules >=3mm**  \
0     LIDC-IDRI-0001                          4                          1   
1     LIDC-IDRI-0002                         12                          1   
2     LIDC-IDRI-0003                          4                          4   
3     LIDC-IDRI-0004                          4                          1   
4     LIDC-IDRI-0005                          9                          3   
...              ...                        ...                        ...   
1014  LIDC-IDRI-1009                          2                          1   
1015  LIDC-IDRI-1010                         10                          1   
1016  LIDC-IDRI-1011                          4                          4   
1017  LIDC-IDRI-1012                          1                          1   
1018             NaN                       7371                       2669   

      Number of Nodules <3mm***  Unnamed: 4  \
0                             3         NaN   
1                            11         NaN   
2                             0         NaN   
3                             3         NaN   
4                             6         NaN   
...                         ...         ...   
1014                          1         NaN   
1015                          9         NaN   
1016                          0         NaN   
1017                          0         NaN   
1018                       4702         NaN   

                                             Unnamed: 5  
0                                                   NaN  
1     *total number of lesions that received either ...  
2     **total number of lesions that received a "nod...  
3     ***total number of lesions that received a "no...  
4                                                   NaN  
...                                                 ...  
1014                                                NaN  
1015                                                NaN  
1016                                                NaN  
1017                                                NaN  
1018                                                NaN  

[1019 rows x 6 columns]

#### Nodule Data Verification

In [2938]:
verif_format_ids(df_nodule,'TCIA Patient ID')                              
                                

The format of IDs in column 'TCIA Patient ID' is not the same.
IDs with different formats:
nan


In [2939]:
duplicated(df_nodule,'TCIA Patient ID')

Duplicated IDs in column 'TCIA Patient ID':
['LIDC-IDRI-0132' 'LIDC-IDRI-0151' 'LIDC-IDRI-0315' 'LIDC-IDRI-0332'
 'LIDC-IDRI-0355' 'LIDC-IDRI-0365' 'LIDC-IDRI-0442' 'LIDC-IDRI-0484']


In [2940]:
missing_data(df_nodule)

Missing values in TCIA Patient ID: 1
Missing values in Total Number of Nodules* : 0
Missing values in Number of Nodules >=3mm**: 0
Missing values in Number of Nodules <3mm***: 0
Missing values in Unnamed: 4: 1019
Missing values in Unnamed: 5: 1016


#### Processing Nodule 's Data

In [2941]:
remove_missing_data(df_nodule, 'TCIA Patient ID')

In [2942]:
missing_data(df_nodule)

Missing values in TCIA Patient ID: 0
Missing values in Total Number of Nodules* : 0
Missing values in Number of Nodules >=3mm**: 0
Missing values in Number of Nodules <3mm***: 0
Missing values in Unnamed: 4: 1018
Missing values in Unnamed: 5: 1015


In [2943]:
df_nodule.shape

(1018, 6)

In [2944]:
process_duplicated(df_nodule, 'TCIA Patient ID')

In [2945]:
duplicated(df_nodule,'TCIA Patient ID')

No duplicated IDs found in column 'TCIA Patient ID'.


In [2946]:
print(df_nodule.columns)
print(df_radiologists.columns)
df_nodule.columns = df_nodule.columns.str.strip()
df_radiologists.columns = df_radiologists.columns.str.strip()
#df_nodule = df_nodule[df_nodule['TCIA Patient ID'].isin(df_radiologists['TCIA Patient ID'])]
common_ids = df_radiologists['TCIA Patient ID']
df_nodule = df_nodule[df_nodule['TCIA Patient ID'].isin(common_ids)]


Index(['TCIA Patient ID', 'Total Number of Nodules* ',
       'Number of Nodules >=3mm**', 'Number of Nodules <3mm***', 'Unnamed: 4',
       'Unnamed: 5'],
      dtype='object')
Index(['TCIA Patient ID', 'Diagnosis at the patient', 'Diagnosis Method',
       'Primary tumor site for metastatic disease',
       'Nodule 1 Diagnosis at the Nodule Level',
       'Nodule 1 Diagnosis Method at the Nodule Level',
       'Nodule 2 Diagnosis at the Nodule Level',
       'Nodule 2 Diagnosis Method at the Nodule Level',
       'Nodule 3 Diagnosis at the Nodule Level',
       'Nodule 3 Diagnosis Method at the Nodule Level',
       'Nodule 4 Diagnosis at the Nodule Level',
       'Nodule 4 Diagnosis Method at the Nodule Level',
       'Nodule 5 Diagnosis at the Nodule Level',
       'Nodule 5 Diagnosis Method at the Nodule Level'],
      dtype='object')


In [2947]:
df_nodule

TCIA Patient ID  Total Number of Nodules*  Number of Nodules >=3mm**  \
67    LIDC-IDRI-0068                         7                          6   
70    LIDC-IDRI-0071                         4                          0   
71    LIDC-IDRI-0072                         3                          1   
87    LIDC-IDRI-0088                         6                          3   
89    LIDC-IDRI-0090                         4                          1   
...              ...                       ...                        ...   
970   LIDC-IDRI-0965                         2                          1   
999   LIDC-IDRI-0994                         9                          6   
1007  LIDC-IDRI-1002                         5                          4   
1009  LIDC-IDRI-1004                        10                          4   
1016  LIDC-IDRI-1011                         4                          4   

      Number of Nodules <3mm***  Unnamed: 4 Unnamed: 5  
67                            1         NaN        NaN  
70                            4         NaN        NaN  
71                            2         NaN        NaN  
87                            3         NaN        NaN  
89                            3         NaN        NaN  
...                         ...         ...        ...  
970                           1         NaN        NaN  
999                           3         NaN        NaN  
1007                          1         NaN        NaN  
1009                          6         NaN        NaN  
1016                          0         NaN        NaN  

[130 rows x 6 columns]

In [2948]:
#After drop the values ID that are not in Radiologists data
print("Shape of nodule = " , df_nodule.shape)


Shape of nodule =  (130, 6)


##### Remove the last two columns

In [2949]:
df_nodule= df_nodule.iloc[:, :-2]
df_nodule

TCIA Patient ID  Total Number of Nodules*  Number of Nodules >=3mm**  \
67    LIDC-IDRI-0068                         7                          6   
70    LIDC-IDRI-0071                         4                          0   
71    LIDC-IDRI-0072                         3                          1   
87    LIDC-IDRI-0088                         6                          3   
89    LIDC-IDRI-0090                         4                          1   
...              ...                       ...                        ...   
970   LIDC-IDRI-0965                         2                          1   
999   LIDC-IDRI-0994                         9                          6   
1007  LIDC-IDRI-1002                         5                          4   
1009  LIDC-IDRI-1004                        10                          4   
1016  LIDC-IDRI-1011                         4                          4   

      Number of Nodules <3mm***  
67                            1  
70                            4  
71                            2  
87                            3  
89                            3  
...                         ...  
970                           1  
999                           3  
1007                          1  
1009                          6  
1016                          0  

[130 rows x 4 columns]

###### Discussion: Applying verif_format_ids funtion, it was verificated that the format is not the same for all ids in TCIA Patient ID column, due to the fact there is a few nan, we resolve it by selecting the common ids with the previous dataset. Then with the duplicated function, it was concluded that there are duplicate ids, which its was processed by calculating the mean between values of duplicated ids. About the missing data, the function indicated that there are missing values that were processed using remove_missing_data function. This last function removed all the missing data in the relevant columns

## Last check of the previous datasets

In [2950]:
if verify_different_ids(df_radiologists, df_nodule, 'TCIA Patient ID', 'TCIA Patient ID'):
    print("There are different IDs between the two columns.")
else:
    print("There are no different IDs between the two columns.")


There are no different IDs between the two columns.


In [2951]:
has_null, null_count = check_null_values(df_radiologists)

if has_null:
    print(f"There are {null_count} null values in df_radiologists")
else:
    print("There are no null values in df_radiologists")

There are 1036 null values in df_radiologists


In [2952]:
has_null, null_count = check_null_values(df_nodule)

if has_null:
    print(f"There are {null_count} null values in df_nodule")
else:
    print("There are no null values in df_nodule")

There are no null values in df_nodule


In [2953]:
ids_with_whitespace(df_radiologists, 'TCIA Patient ID')


No IDs with whitespace


In [2954]:
ids_with_whitespace(df_nodule, 'TCIA Patient ID')

No IDs with whitespace


###### There are no different IDs between the two columns; There are 1036 null values in df_radiologists, which we decide not to remove so as not to have a very small number of data. On the other hand the nodule dataset doens´t have any null values. Both datasets don´t have white spaces.

## Load MetaData

In [2955]:
import pandas as pd

# Specify the file path
file_path = '/Users/larasousa/Desktop/uni/MetaData.csv'

# Load the dataset into a pandas DataFrame
df_metadata = pd.read_csv(file_path)

# Now 'data' contains your dataset, and you can perform various operations on it.
df_metadata

Subject ID                                          Study UID  \
0     LIDC-IDRI-1001  1.3.6.1.4.1.14519.5.2.1.6279.6001.281499745765...   
1     LIDC-IDRI-0778  1.3.6.1.4.1.14519.5.2.1.6279.6001.174809695196...   
2     LIDC-IDRI-0813  1.3.6.1.4.1.14519.5.2.1.6279.6001.139110171863...   
3     LIDC-IDRI-0710  1.3.6.1.4.1.14519.5.2.1.6279.6001.116146223752...   
4     LIDC-IDRI-0410  1.3.6.1.4.1.14519.5.2.1.6279.6001.818775038273...   
...              ...                                                ...   
1303  LIDC-IDRI-0030  1.3.6.1.4.1.14519.5.2.1.6279.6001.573428694448...   
1304  LIDC-IDRI-0121  1.3.6.1.4.1.14519.5.2.1.6279.6001.985098410443...   
1305  LIDC-IDRI-0974  1.3.6.1.4.1.14519.5.2.1.6279.6001.214565862082...   
1306  LIDC-IDRI-0473  1.3.6.1.4.1.14519.5.2.1.6279.6001.210105060472...   
1307  LIDC-IDRI-0493  1.3.6.1.4.1.14519.5.2.1.6279.6001.333362756208...   

     Study Description             Study Date  \
0                  NaN  2000-01-01 00:00:00.0   
1                CHEST  2000-01-01 00:00:00.0   
2                  NaN  2000-01-01 00:00:00.0   
3                CHEST  2000-01-01 00:00:00.0   
4       CT LUNG SCREEN  2000-01-01 00:00:00.0   
...                ...                    ...   
1303               NaN  2000-01-01 00:00:00.0   
1304               NaN  2000-01-01 00:00:00.0   
1305               NaN  2000-01-01 00:00:00.0   
1306  CT CHEST O CONTR  2000-01-01 00:00:00.0   
1307             CHEST  2000-01-01 00:00:00.0   

                                              Series ID    Series Description  \
0     1.3.6.1.4.1.14519.5.2.1.6279.6001.100225287222...                   NaN   
1     1.3.6.1.4.1.14519.5.2.1.6279.6001.100332161840...              Recon 2:   
2     1.3.6.1.4.1.14519.5.2.1.6279.6001.100398138793...                   NaN   
3     1.3.6.1.4.1.14519.5.2.1.6279.6001.100530488926...  Recon 2: ACRIN LARGE   
4     1.3.6.1.4.1.14519.5.2.1.6279.6001.100620385482...                   NaN   
...                                                 ...                   ...   
1303  1.3.6.1.4.1.14519.5.2.1.6279.6001.986011151772...                   NaN   
1304  1.3.6.1.4.1.14519.5.2.1.6279.6001.987704869630...                   NaN   
1305  1.3.6.1.4.1.14519.5.2.1.6279.6001.991510467496...                   NaN   
1306  1.3.6.1.4.1.14519.5.2.1.6279.6001.994459772950...                   NaN   
1307  1.3.6.1.4.1.14519.5.2.1.6279.6001.997611074084...                   NaN   

      Number of images  File Size (Bytes) Collection Name Modality  \
0                  194          102096016       LIDC-IDRI       CT   
1                  481          253245158       LIDC-IDRI       CT   
2                  157           82630786       LIDC-IDRI       CT   
3                  471          247987448       LIDC-IDRI       CT   
4                  126           66314722       LIDC-IDRI       CT   
...                ...                ...             ...      ...   
1303               119           62645518       LIDC-IDRI       CT   
1304                 1           13228740       LIDC-IDRI       CR   
1305               101           53158868       LIDC-IDRI       CT   
1306               291          153155158       LIDC-IDRI       CT   
1307               285          150013636       LIDC-IDRI       CT   

                 Manufacturer  
0                     SIEMENS  
1          GE MEDICAL SYSTEMS  
2                     SIEMENS  
3          GE MEDICAL SYSTEMS  
4          GE MEDICAL SYSTEMS  
...                       ...  
1303       GE MEDICAL SYSTEMS  
1304  Philips Medical Systems  
1305                  SIEMENS  
1306       GE MEDICAL SYSTEMS  
1307       GE MEDICAL SYSTEMS  

[1308 rows x 11 columns]

### MetaData verification

In [2956]:
verif_format_ids(df_metadata, 'Subject ID')

The format of IDs in column 'Subject ID' is the same (LIDC-IDRI- followed by a four-digit number).


In [2957]:
duplicated(df_metadata, 'Subject ID')

Duplicated IDs in column 'Subject ID':
['LIDC-IDRI-0159' 'LIDC-IDRI-0182' 'LIDC-IDRI-0026' 'LIDC-IDRI-0252'
 'LIDC-IDRI-0249' 'LIDC-IDRI-0187' 'LIDC-IDRI-0245' 'LIDC-IDRI-0134'
 'LIDC-IDRI-0123' 'LIDC-IDRI-0066' 'LIDC-IDRI-0138' 'LIDC-IDRI-0114'
 'LIDC-IDRI-0062' 'LIDC-IDRI-0236' 'LIDC-IDRI-0278' 'LIDC-IDRI-0184'
 'LIDC-IDRI-0122' 'LIDC-IDRI-0017' 'LIDC-IDRI-0225' 'LIDC-IDRI-0165'
 'LIDC-IDRI-0285' 'LIDC-IDRI-0286' 'LIDC-IDRI-0124' 'LIDC-IDRI-0111'
 'LIDC-IDRI-0088' 'LIDC-IDRI-0202' 'LIDC-IDRI-0060' 'LIDC-IDRI-0135'
 'LIDC-IDRI-0030' 'LIDC-IDRI-0095' 'LIDC-IDRI-0154' 'LIDC-IDRI-0096'
 'LIDC-IDRI-0042' 'LIDC-IDRI-0038' 'LIDC-IDRI-0144' 'LIDC-IDRI-0166'
 'LIDC-IDRI-0032' 'LIDC-IDRI-0260' 'LIDC-IDRI-0229' 'LIDC-IDRI-0207'
 'LIDC-IDRI-0046' 'LIDC-IDRI-0226' 'LIDC-IDRI-0031' 'LIDC-IDRI-0108'
 'LIDC-IDRI-0039' 'LIDC-IDRI-0043' 'LIDC-IDRI-0052' 'LIDC-IDRI-0044'
 'LIDC-IDRI-0005' 'LIDC-IDRI-0033' 'LIDC-IDRI-0262' 'LIDC-IDRI-0059'
 'LIDC-IDRI-0101' 'LIDC-IDRI-0086' 'LIDC-IDRI-0148' 'LIDC-IDRI-0

In [2958]:
missing_data(df_metadata)

Missing values in Subject ID: 0
Missing values in Study UID: 0
Missing values in Study Description: 691
Missing values in Study Date: 0
Missing values in Series ID: 0
Missing values in Series Description: 844
Missing values in Number of images: 0
Missing values in File Size (Bytes): 0
Missing values in Collection Name: 0
Missing values in Modality: 0
Missing values in Manufacturer: 0


In [2959]:
has_null, null_count = check_null_values(df_metadata)

if has_null:
    print(f"There are {null_count} null values in metadata")
else:
    print("There are no null values in metadata")


There are 1535 null values in metadata
